### PR TITLE
Make layout compatible with iPhone X

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
This PR makes the whole app layout compatible with iPhone, iPhone X and Android

I think I made all the screens work as expected but testing through all devices might be something you should do when reviewing

#### Ticket Link
https://mattermost.atlassian.net/browse/RN-530

### CODE REVIEW

**Comment by hmhealey** [comment]
I just tested on Nexus 5 on Android 6.0, and the only thing I noticed compared to 1.4 is that the background on the Jump to Conversation box is missing, but that might not be related to your PR
image

File: app/components/client_upgrade_listener/client_upgrade_listener.js [hunk 1]
```diff
@@ -110,27 +140,39 @@ class ClientUpgradeListener extends PureComponent {
         });
 
         this.toggleUpgradeMessage(false);
-    }
+    };
 
     handleLearnMore = () => {
-        this.props.navigator.dismissAllModals({animationType: 'none'});
+        const {intl} = this.context;
+        this.props.navigator.dismissModal({animationType: 'none'});
```
Comment (hmhealey) [inline app/components/client_upgrade_listener/client_upgrade_listener.js]: Did you mean to change this from dismissAllModals to dismissModal?

File: app/components/emoji_picker/emoji_picker.js [hunk 2]
```diff
@@ -74,6 +77,10 @@ export default class EmojiPicker extends PureComponent {
             this.setState({
                 emojis: this.renderableEmojis(this.props.emojisBySection, nextProps.deviceWidth)
             });
+
+            if (this.refs.search_bar) {
+                this.refs.search_bar.blur();
```
Comment (hmhealey) [inline app/components/emoji_picker/emoji_picker.js]: Why do we need to blur the search bar when rotating the device?

File: app/components/safe_area_view/safe_area_view.android.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import {View} from 'react-native';
+
+export default class SafeAreaAndroid extends PureComponent {
+    static propTypes = {
+        children: PropTypes.node.isRequired
+    };
+
+    render() {
+        return (
+            <View style={{flex: 1}}>
```
Comment (hmhealey) [inline app/components/safe_area_view/safe_area_view.android.js]: Can you switch to a StyleSheet here?

**Comment by enahum** [comment]
Actually is not, i’m submitting another PR in a few minutes that fixes exactly that problem

File: app/components/safe_area_view/safe_area_view.ios.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    keyboardWillShow = () => {
+        this.setState({keyboard: true});
+    };
+
+    onNavigatorEvent = (event) => {
+        switch (event.id) {
+        case 'willAppear':
+        case 'didDisappear':
+            this.getSafeAreaInsets();
+            Keyboard.dismiss();
```
Comment (hmhealey) [inline app/components/safe_area_view/safe_area_view.ios.js]: This doesn't look right to me. Why do we need to dismiss the keyboard here? It sounds like we're trying to dismiss it either when it hasn't opened yet (but is opening) or when it's already closed

File: app/screens/search/search.js [hunk 5]
```diff
@@ -87,7 +91,9 @@ class Search extends PureComponent {
     componentDidMount() {
         this.setManagedConfig();
         if (this.refs.searchBar) {
-            this.refs.searchBar.focus();
+            setTimeout(() => {
+                this.refs.searchBar.focus();
+            }, 500);
```
Comment (hmhealey) [inline app/screens/search/search.js]: Is it possible to reduce this a bit more to around maybe 150ms? That should keep this feeling snappy

File: app/components/client_upgrade_listener/client_upgrade_listener.js (hunk 1 above)
Comment (enahum) [reply app/components/client_upgrade_listener/client_upgrade_listener.js]: Yes

File: app/components/emoji_picker/emoji_picker.js (hunk 2 above)
Comment (enahum) [reply app/components/emoji_picker/emoji_picker.js]: it actually blurs but does not remove the `Cancel` button which makes it behave erroneously, by doing this we avoid the erroneous behavior

File: app/components/safe_area_view/safe_area_view.ios.js (hunk 4 above)
Comment (enahum) [reply app/components/safe_area_view/safe_area_view.ios.js]: yeap, something that I forgot to remove

File: app/screens/search/search.js (hunk 5 above)
Comment (enahum) [reply app/screens/search/search.js]: I'll play with this value, but this is the best I came up with and it works every time, and if you tested it, you can see is really smooth

**Comment by mkraft** [comment]
I rebased this onto master locally and can't get it to run on either the iPhone 6 or iPhone X emulators, I get the following:

simulator screen shot - iphone x - 2017-11-27 at 18 07 32

**Comment by enahum** [comment]
Try updating mattermost-redux

**Comment by mkraft** [comment]
I actually had to downgrade. I believe https://github.com/mattermost/mattermost-redux/commit/fb083f67b5560ff64fe251415460a1af5d690e5a breaks something. Anyhow I believe it is not related to this PR.

**Comment by mkraft** [comment]
That is strange. I did a `yarn upgrade mattermost-redux`, reset cache etc etc. Didn't work. Then I changed from `mattermost/mattermost-redux#master` to `mattermost/mattermost-redux#fb083f6` and back to master and then it started working. Something got into an invalid state.

**Comment by enahum** [comment]
Yeah but it started to happen with that commit, I think we should revert it, I’ll have a look tomorrow at that commit and see what can be the cause or have a discussion around it

**Pull request merged**